packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp

```diff
@@ -0,0 +1,133 @@
+ *  This class provides a means to modify just the solution values
+ *  (i.e., \f$x\f$ and \f$dot{x}\f$), and nothing else, but time and
+ *  timestep are also provided.
+ *
+ *  Users deriving from this class can access and change the solution
+ *  during the timestep (e.g., limiting the solution for monoticity).
+ *  It is expected that the user knows what changes are allowable without
+ *  affecting the Stepper correctness, performance, accuracy and stability
+ *  (i.e., USER BEWARE!!).
+ *
+ *  Below is the BDF2 algorithm with the locations of the ModifierX calls
+ *  italicized.
+ *
+ *  \f{algorithm}{
+ *  \renewcommand{\thealgorithm}{}
+ *  \caption{BDF2 with the locations of the application actions indicated}
+ *  \begin{algorithmic}[1]
+ *    \State Start with $x_n$, $\Delta t_n$
+ *    \State {\it appAction.execute(solutionHistory, stepper, BEGIN\_STEP)}
+ *    \State Form $f(x_{n},t_{n})$
+ *    \State {\it appAction.execute(solutionHistory, stepper, BEFORE\_EXPLICIT\_EVAL)}
+ *    \State Form $x_n \leftarrow x_{n} + \Delta t_n f(x_{n},t_n)$
+ *    \State {\it appAction.execute(solutionHistory, stepper, END\_STEP)}
+ *  \end{algorithmic}
+ *  \f}
+ */
+
+template<class Scalar>
+class StepperBDF2ModifierXBase
+  : virtual public Tempus::StepperBDF2AppAction<Scalar>
+{
+private:
+
+  /* \brief Adaptor execute function
+   *
+   *  This is an adaptor function to bridge between the AppAction
+   *  interface and the ModifierX interface.  It is meant to be private
+   *  and non-virtual as deriving from this class should only need to
+   *  implement the modify function.
+   *
+   *  For the ModifierX interface, this adaptor maps the
+   *  StepperBDF2AppAction::ACTION_LOCATION to the
+   *  StepperBDF2ModifierX::MODIFIERX_TYPE, and only pass the solution
+   *  (\f$x\f$ and/or \f$\dot{x}\f$ and other parameters to the modify
+   *  function.
+   */
+  void execute(
+    Teuchos::RCP<SolutionHistory<Scalar> > sh,
+    Teuchos::RCP<StepperBDF2<Scalar> > stepper,
+    const typename StepperBDF2AppAction<Scalar>::ACTION_LOCATION actLoc)
+  {
+    using Teuchos::RCP;
+
+    MODIFIER_TYPE modType = X_BEGIN_STEP;
+    RCP<SolutionState<Scalar> > workingState = sh->getWorkingState();
+    const Scalar time = workingState->getTime();
+    const Scalar dt   = workingState->getTimeStep();
+    RCP<Thyra::VectorBase<Scalar> > x;
+
+    switch(actLoc) {
+      case StepperBDF2AppAction<Scalar>::BEGIN_STEP:
+      {
+        modType = X_BEGIN_STEP;
+        x = workingState->getX();
+        break;
+      }
+      case StepperBDF2AppAction<Scalar>::BEFORE_SOLVE:
+      {
+        modType = X_BEFORE_SOLVE;
+        x = workingState->getX();
+        break;
+      }
+      case StepperBDF2AppAction<Scalar>::AFTER_SOLVE:
+      {
+        modType = X_AFTER_SOLVE;
+        x = workingState->getX();
+        break;
+      }
+      case StepperBDF2AppAction<Scalar>::END_STEP:
+      {
+        modType = XDOT_END_STEP;
+        x = stepper->getStepperXDot();
```

Comment:
We should make this X_END_STEP and not XDOT_END_STEP, using getStepperX().

Reply:
@ccober6 So no getXdots anywhere for BDF2?

Reply:
Yeah, in talking with the EMPIRE folks they leaned towards just accessing X and not XDot.